Automattic/jetpack
# Boost: Migrate Critical CSS related components to React

## Proposed changes:
* This PR updates the two main Cloud CSS and Critical CSS components (and their children) to React.

### Other information:


## Jetpack product discussion

pc9hqz-2jy-p2

## Does this pull request change what data or activity we track or use?

no

## Testing instructions:

### Testing that Critical CSS generation works

<details><summary>Click to see image</summary><img src="https://github.com/Automattic/jetpack/assets/11799079/7d22de7a-24ef-4c52-8af6-315ba46a50af" /></details>

* Setup the free version of Boost;
* Generate Critical CSS;
* Make sure the progress bar, regenerate button and progress text work accordingly.

### Testing that Critical CSS generation shows something went wrong with some of the pages

<details><summary>Click to see image</summary><img src="https://github.com/Automattic/jetpack/assets/11799079/356d9358-a476-4a20-b7f4-bb761daf8c6d" /></details>

* Add the following code to your theme (or as an mu plugin):

```php
<?php

add_filter( 'rest_pre_dispatch', function ( $result, $server, $request ) {
	if ( '/jetpack-boost/v1/critical-css/start' === $request->get_route() ) {
		update_option( 'test_errors', 0 );
	}

	return $result;
}, 10, 3 );

add_action( 'template_redirect', function () {
	$test_errors = get_option( 'test_errors', 0 );
	$test_errors++;
	update_option( 'test_errors', $test_errors );

	if ( $test_errors >= 4 ) {
		nonexistentfunction();
	}
} );

```
* Regenerate critical CSS;
* This will cause an error 500 for **some** pages and you should see the UI;
* You can remove the code and click Regenerate to make sure everything works fine after that.

### Testing that Critical CSS generation shows the "Failed to generate..." error

<details><summary>Click to see image</summary><img src="https://github.com/Automattic/jetpack/assets/11799079/914c1629-dd50-48e7-97d3-8d4744dc6052" /></details>

* Add the following code to your theme (or as an mu plugin):

```php
<?php

add_action( 'template_redirect', function () {
	if ( ! isset( $_GET['jb-generate-critical-css'] ) ) {
		return;
	}

	noexistentfunction();
} );
```
* Regenerate critical CSS;
* This will cause an error 500 for **all** pages and you should see the UI;
* Click the **See error message** link to make sure the detailed error shows up;
* You can remove the code and click Regenerate to make sure everything works fine after that.

### Testing that Cloud CSS works properly
 
* Setup the paid version of Boost and make sure the UI for generating critical css works as expected/before 😄

### Testing that Cloud CSS generation shows something went wrong with some of the pages

<details><summary>Click to see image</summary><img src="https://github.com/Automattic/jetpack/assets/11799079/bed83641-964f-40fd-8c61-22193e8aceb8" /></details>

* Add the following code to your theme (or as an mu plugin):

```php
<?php

add_action( 'template_redirect', function () {
	if ( ! is_front_page() ) {
		noexistentfunction();
	}
} );

```
* Regenerate critical CSS;
* This will cause an error 500 for every page other than the home page;
* Remove the snippet and click regenerate to make sure the UI works.

### Testing that Cloud CSS generation shows its various error UI stages correctly

Honestly, I couldn't figure out how to normally cause these to show up, so I just update the state value in the DB manually, to force the UI to show up. I'm using the following script with different values based on which UI version I want to see:

https://gist.github.com/dilirity/75bed82de1c10e3dd26d6ee1ccf4412d

Just open the home page with the `jb-set-cloud-css-state` GET param and use one of the following values:

- [show-critical-css-show-stopper-simple](http://jetpack-boost.test/?jb-set-cloud-css-state=show-critical-css-show-stopper-simple)

<details><summary>Click to see image</summary><img src="https://github.com/Automattic/jetpack/assets/11799079/6a13d3f3-e4fd-4ed1-8ed1-7d7edb8f7cbe" /></details>

- [show-advanced-css-recommendations](http://jetpack-boost.test/?jb-set-cloud-css-state=show-advanced-css-recommendations)

<details><summary>Click to see image</summary><img src="https://github.com/Automattic/jetpack/assets/11799079/531c22f4-1c3e-49a5-8752-b4772f251846" /></details>

- [show-critical-css-show-stopper-detailed](http://jetpack-boost.test/?jb-set-cloud-css-state=show-critical-css-show-stopper-detailed)

<details><summary>Click to see image</summary><img src="https://github.com/Automattic/jetpack/assets/11799079/d29abd71-0ecc-47a6-8d65-c812d802ddf2" /></details>
